File: src/fundamentals/conformance.rs
use crate::*;
use std::rc::Rc;

#[derive(Clone)]
pub struct Conformance<A: 'static + Clone, B: 'static + Clone> {
    pub derived_trait_id: TraitID<B>,
    pub validation: Validation<Value, A>,
    pub derive_trait_value: Rc<dyn Fn(A, &mut Environment, &ProgramStack) -> Result<B>>,
}

impl<A: Clone, B: Clone> Conformance<A, B> {
    pub fn new(
        derived_trait_id: TraitID<B>,
        validation: Validation<Value, A>,
        derive_trait_value: impl Fn(A, &mut Environment, &ProgramStack) -> Result<B> + 'static,
    ) -> Conformance<A, B> {
        Conformance {
            derived_trait_id,
            validation,
            derive_trait_value: Rc::new(derive_trait_value),
        }
    }
}

pub type AnyConformance = Conformance<Any, Any>;

impl AnyConformance {
    fn from<A: Clone, B: Clone>(conformance: Conformance<A, B>) -> AnyConformance {
        AnyConformance {
            derived_trait_id: AnyTraitID::from(conformance.derived_trait_id.clone()),
            validation: Validation::new({
                let conformance = conformance.clone();

                move |value, env, stack| {
                    let erased_result: ValidationResult<Any> =
                        match conformance.validation.validate(value, env, stack)? {
                            Valid(value) => Valid(Any::from(value)),
                            Invalid => Invalid,
                        };

                    Ok(erased_result)
                }
            }),
            derive_trait_value: Rc::new(move |value, env, stack| {
                (conformance.derive_trait_value)(value.cast::<A>().clone(), env, &stack)
                    .map(|v| Any::from(v))
            }),
        }
    }
}

impl<A: Clone, B: Clone> Conformance<A, B> {
    pub fn from_any(conformance: AnyConformance) -> Conformance<A, B> {
        Conformance::new(
            TraitID::from_any(conformance.clone().derived_trait_id),
            Validation::new({
                let conformance = conformance.clone();

                move |value, env, stack| {
                    let value = match conformance.validation.validate(value, env, stack)? {
                        Valid(erased_value) => {
                            let value = erased_value.cast::<A>().clone();
                            Valid(value)
                        }
                        Invalid => Invalid,
                    };

                    Ok(value)
                }
            }),
            move |value, env, stack| {
                let erased_value =
                    (conformance.derive_trait_value)(Any::from(value.clone()), env, stack)?;
                let value = erased_value.cast::<B>().clone();
                Ok(value)
            },
        )
    }
}

impl Environment {
    pub fn add_conformance<A: Clone, B: Clone>(&mut self, conformance: Conformance<A, B>) {
        self.conformances.push(AnyConformance::from(conformance));
    }
}

File: Sources/Fundamentals/Conformance.swift
public struct Conformance<A, B> {
    public var derivedTraitID: TraitID<B>
    public var validation: Validation<Value, A>
    public var deriveTraitValue: (A, inout Environment, ProgramStack) throws -> B

    public init(
        derivedTraitID: TraitID<B>,
        validation: Validation<Value, A>,
        deriveTraitValue: @escaping (A, inout Environment, ProgramStack) throws -> B
    ) {
        self.derivedTraitID = derivedTraitID
        self.validation = validation
        self.deriveTraitValue = deriveTraitValue
    }
}

public typealias AnyConformance = Conformance<Any, Any>

extension AnyConformance {
    public init<A, B>(
        _ conformance: Conformance<A, B>
    ) {
        self
            .init(
                derivedTraitID: AnyTraitID(conformance.derivedTraitID),
                validation: Validation(debugLabel: conformance.validation.debugLabel) { value, env, stack in
                    switch try conformance.validation(value, &env, stack) {
                    case .valid(let value):
                        return .valid(value)
                    case .invalid:
                        return .invalid
                    }
                },
                deriveTraitValue: { value, env, stack in
                    try conformance.deriveTraitValue(value as! A, &env, stack)
                }
            )
    }
}

extension Conformance {
    public init(
        fromAny conformance: AnyConformance
    ) {
        self.init(
            derivedTraitID: TraitID(fromAny: conformance.derivedTraitID),
            validation: Validation(debugLabel: conformance.validation.debugLabel) { value, env, stack in
                switch try conformance.validation(value, &env, stack) {
                case .valid(let value):
                    return .valid(value as! A)
                case .invalid:
                    return .invalid
                }
            },
            deriveTraitValue: { value, env, stack in
                try conformance.deriveTraitValue(value, &env, stack) as! B
            }
        )
    }
}

extension Environment {
    public mutating func addConformance<A, B>(_ conformance: Conformance<A, B>) {
        self.conformances.append(AnyConformance(conformance))
    }
}

extension Conformance: CustomDebugStringConvertible {
    public var debugDescription: String {
        "\(self.validation.debugLabel ?? "<validation>") ::= \(self.derivedTraitID.debugLabel ?? "<trait>")"
    }
}
